aim/db/hashtree_db_listener.py

```diff
@@ -253,8 +255,29 @@ def _preprocess_logs(self, ctx, logs):
                     if action != aim_tree.ActionLog.DELETE:
                         LOG.warn("AIM resource %s does not exist in DB "
                                  "for create/update action" % aim_res)
+                        action = aim_tree.ActionLog.SKIP
+
+                # Queue up these SG rules first as we really just need
+                # the last one instead of sending all those duplicate
+                # rules over while building the tree.
+                if action != aim_tree.ActionLog.SKIP:
+                    sg_rule_logs.setdefault(aim_res.name, []).append(
+                        [action, aim_res, log])
+                    continue
             log_by_root.setdefault(log.root_rn, []).append(
                 (action, aim_res, log))
+
+        for sg_rule_name, list_of_logs in sg_rule_logs.items():
+            # We will skip the processing of all the sg_rules except
+            # for the last one.
+            for index, items in enumerate(list_of_logs[:-1]):
+                items[0] = aim_tree.ActionLog.SKIP
+                list_of_logs[index] = tuple(items)
+            # Convert the last one also.
```

Comment:
The intent of lines 240-253 were to just use what's actually in the DB, rather than iterate through all the actions, with the idea being that the DB (or at least the DB within this transaction) has the desired end state. Should the loop instead just be looking for the last entry with the matching state (I don't know how/why it won't be the last one, but it might be worth checking)?

Reply:
The last one could also be a DELETE operation though. Its true that we will only need the last one however we still need to pass the whole list to the log_by_root dictionary because that's the only way we can guarantee those action logs got deleted later on here,
https://github.com/noironetworks/aci-integration-module/blob/master/aim/db/hashtree_db_listener.py#L319

But I think I don't have to create another loop here at line#273 because I can mark those entries as SKIP while queuing up those SG rules related action logs around line#264. Let me do that then, which can speed things up.